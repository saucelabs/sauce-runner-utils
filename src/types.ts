export type Suite = {
  name: string;
  env?: { [key: string]: string };
  config?: {
    env?: { [key: string]: string };
  };
};

export type Registry = {
  scope?: string;
  url: string;
  authToken?: string;
  auth?: string;
  username?: string;
  password?: string;
  email?: string;
};

export type NpmConfig = {
  /**
   * @deprecated: registry should be avoided in favor of a "registries" entry.
   */
  registry?: string;
  registries?: Registry[];
  strictSSL?: boolean | null;
  packageLock?: boolean | string | null;
  packages?: { [key: string]: string | number };
  legacyPeerDeps?: boolean | string | null;
};

export interface PathContainer {
  path: string;
}

export interface NpmConfigContainer {
  npm?: NpmConfig;
}

export interface SuitesContainer {
  suites: Suite[];
}

export interface NodeContext {
  nodePath: string;
  npmPath: string;

  // Ignore the specific binary paths provided and rely on the binaries
  // from PATH.
  useGlobals?: boolean;
}
