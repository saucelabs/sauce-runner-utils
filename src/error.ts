export function errCode(obj: unknown): string | number {
  if (typeof obj !== 'object' || obj === null || !('code' in obj)) {
    return '';
  }
  if (typeof obj.code === 'number' || typeof obj.code === 'string') {
    return obj.code;
  }
  return '';
}
