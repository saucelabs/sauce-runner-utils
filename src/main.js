module.exports = require('./utils');
module.exports.saucectl = require('./saucectl');