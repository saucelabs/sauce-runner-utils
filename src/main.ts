/* istanbul ignore file */
// import utils from './utils';
// import saucectl from './saucectl';

// export default import from './utils';
// export saucectl { default as saucectl } from './saucectl';
