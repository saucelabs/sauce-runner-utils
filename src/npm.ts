import { spawn } from 'child_process';
import { lstat, rename, rm, writeFile } from 'fs/promises';
import { IHasNodePath } from './types';

const temporarilyMovedFiles: {[key: string]: string} = {
  'package.json': `.package.json-${process.pid}`,
  'package-lock.json': `.package-lock.json-${process.pid}`,
};

export default class NPM {
  static async removePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(name);
        await rename(name, replacement);
      } catch (e) {
        console.debug(`no ${name} to archive`);
      }
    }
  }

  static async restorePackageJson () {
    for (const [name, replacement] of Object.entries(temporarilyMovedFiles)) {
      try {
        // Note: we don't need the return value.
        // lstat is only used to check that the file exists.
        await lstat(replacement);
        await rename(replacement, name);
      } catch (e) {
        console.debug(`no ${name} was archived`);
      }
    }
  }

  public static async install (nodePath: IHasNodePath, pkg: {[key: string]: string}) {
    await this.removePackageJson();
    await writeFile('package.json', JSON.stringify({
      dependencies: pkg,
    }));

    const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'install']);
    p.stdout.pipe(process.stdout);
    p.stderr.pipe(process.stderr);

    const exitPromise = new Promise((resolve) => {
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });

    const exitCode = await exitPromise;

    await rm('package.json', { force: true });
    await this.restorePackageJson();

    return exitCode;
  }

  public static configure (nodePath: IHasNodePath, cfg: {[key: string]: object | string | number | boolean | null }): Promise<number | null> {
    return new Promise((resolve) => {
      const args = Object.keys(cfg).map((k,) => `${k}=${cfg[k]}`);
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'config', 'set', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', () => {
        console.log('Finished');
        resolve(0);
      });
    });
  }

  public static rebuild (nodePath: IHasNodePath, ...args: string[]): Promise<number | null> {
    return new Promise((resolve) => {
      const p = spawn(nodePath.nodePath, [nodePath.npmPath, 'rebuild', ...args]);
      p.stdout.pipe(process.stdout);
      p.stderr.pipe(process.stderr);
      p.on('exit', (exitCode) => {
        resolve(exitCode);
      });
    });
  }
}

