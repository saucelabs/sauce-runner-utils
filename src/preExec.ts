import ChildProcess from 'child_process';
import os from 'os';

function spawnAsync (cmd: string, args: string[]) {
  return new Promise(function (resolve) {
    const proc = ChildProcess.spawn(cmd, args);
    proc.stdout.on('data', (data) => {
      process.stdout.write(data.toString());
    });
    proc.stderr.on('data', (data) => {
      process.stderr.write(data.toString());
    });
    proc.on('exit', function (exitCode) {
      resolve(exitCode);
    });
    proc.on('error', function (err) {
      console.log(`Unable to start command: ${err}`);
      resolve(1);
    });
  });
}

async function preExecRunner (preExecs: string[]) {
  const cmdInvoker = (os.platform() === 'win32') ? 'cmd' : 'sh';
  const cmdArg = (os.platform() === 'win32') ? '/C' : '-c';

  for (const command of preExecs) {
    console.log(`Executing pre-exec command: ${command}`);
    const exitCode = await spawnAsync(cmdInvoker, [cmdArg, command]);
    console.log('\n');

    if (exitCode !== 0) {
      return false;
    }
  }
  return true;
}

export async function run (suite: { preExec: string[] | undefined | null }, timeoutSec: number): Promise<boolean> {
  if (!suite.preExec) {
    return true;
  }

  let timeout;
  const timeoutPromise = new Promise((resolve) => {
    timeout = setTimeout(() => {
      console.error(`Pre-Exec timed out after ${timeoutSec} seconds`);
      resolve(false);
    }, timeoutSec * 1000);
  });
  const hasPassed: boolean = await Promise.race([timeoutPromise, preExecRunner(suite.preExec)]) as boolean;
  clearTimeout(timeout);
  return hasPassed;
}

export default run;
