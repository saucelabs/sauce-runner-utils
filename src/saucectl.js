const fs = require('fs');

// Store file containing job-details url.
// Path has to match the value of the Dockerfile label com.saucelabs.job-info !
const OUTPUT_FILE_PATH = '/tmp/output.json';

function exportValue (payload) {
  fs.writeFileSync(OUTPUT_FILE_PATH, JSON.stringify(payload));
}

function updateExportedValue (data) {
  let fileData;
  try {
    const st = fs.statSync(OUTPUT_FILE_PATH);
    if (st.isFile()) {
      fileData = JSON.parse(fs.readFileSync(OUTPUT_FILE_PATH)) || {};
    }
  } catch (e) {}
  fileData = { ...fileData, ...data };
  exportValue(fileData);
}

module.exports = {
  OUTPUT_FILE_PATH, exportValue, updateExportedValue,
};