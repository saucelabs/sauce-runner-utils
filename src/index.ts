import * as saucectl from './saucectl';
import * as npm from './npm';
import * as utils from './utils';
import {
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
} from './utils';

export {
  npm,
  saucectl,
  utils,

  // Exporting all to keep compatibility with previous API
  getAbsolutePath,
  shouldRecordVideo,
  loadRunConfig,
  getDefaultRegistry,
  setUpNpmConfig,
  installNpmDependencies,
  rebuildNpmDependencies,
  hasNodeModulesFolder,
  getNpmConfig,
  prepareNpmEnv,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  escapeXML
};
