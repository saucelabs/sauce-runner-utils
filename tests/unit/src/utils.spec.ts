// vim: tabstop=2 shiftwidth=2 expandtab
jest.mock('child_process');
jest.mock('fs');
jest.mock('../../../src/npm');

import path from 'path';
import fs from 'fs';
import {
  getAbsolutePath,
  shouldRecordVideo,
  getArgs,
  getEnv,
  getSuite,
  renameScreenshot,
  renameAsset,
  prepareNpmEnv,
  setUpNpmConfig,
  getNpmConfig,
  installNpmDependencies,
  escapeXML } from '../../../src/utils';
import _ from 'lodash';
import npm from '../../../src/npm';
import { NodeContext, NpmConfigContainer, PathContainer, SuitesContainer, Suite } from '../../../src/types';

describe('utils', function () {
  const nodePath: NodeContext = { nodePath: 'node-bin', npmPath: 'npm-bin' };

  describe('.getNpmConfig', function () {
    const emptyConfig = {
      npm: {}
    };

    it('should return empty dict when runner config is empty', function () {
      const npmConfig = getNpmConfig({});

      expect(npmConfig).toStrictEqual({});
    });

    it('should set values when runner npm config is empty', function () {
      const npmConfig = getNpmConfig(emptyConfig);

      expect(npmConfig).toHaveProperty('strict-ssl');
      expect(npmConfig).toHaveProperty('registry');
      expect(npmConfig).toHaveProperty('package-lock');
    });

    it('should set strictSSL to true by default', function () {
      const npmConfig = getNpmConfig(emptyConfig);
      expect(npmConfig).toHaveProperty('strict-ssl', true);
    });

    it('should set strictSSL from runner config', function () {
      const runnerConfig: NpmConfigContainer = {
        npm: {
          strictSSL: false
        }
      };
      let npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('strict-ssl', false);

      runnerConfig.npm ||= {};
      runnerConfig.npm.strictSSL = true;
      npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('strict-ssl', true);

      runnerConfig.npm.strictSSL = 'truthy?';
      npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('strict-ssl', true);
    });

    it('should set packageLock to false by default', function () {
      const npmConfig = getNpmConfig(emptyConfig);

      expect(npmConfig).toHaveProperty('package-lock', false);
    });

    it('should set packageLock from runner config', function () {
      const runnerConfig: NpmConfigContainer = {
        npm: {
          'packageLock': true
        }
      };
      let npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('package-lock', true);

      runnerConfig.npm ||= {};
      runnerConfig.npm.packageLock = false;
      npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('package-lock', false);

      runnerConfig.npm.packageLock = 'truthy?';
      npmConfig = getNpmConfig(runnerConfig);
      expect(npmConfig).toHaveProperty('package-lock', false);
    });

    it('should set the default npm registry by default', function () {
      const npmConfig = getNpmConfig(emptyConfig);

      expect(npmConfig).toHaveProperty('registry');
      expect(npmConfig.registry).not.toBe('');
    });

    it('should set registry from runner config', function () {
      const runnerConfig = {
        npm: {
          registry: 'http://my-private-registry.com'
        }
      };
      const npmConfig = getNpmConfig(runnerConfig);

      expect(npmConfig).toHaveProperty('registry', 'http://my-private-registry.com');
    });
  });

  describe('.prepareNpmEnv', function () {
    let backupEnv: {[key: string]: string | undefined};
    const runCfg: NpmConfigContainer & PathContainer = {
      path: '/fake/runner/path',
      npm: {
        packages: {
          'left-pad': '1.3.0',
          jquery: 2,
        }
      }
    };
    beforeEach(function () {
      backupEnv = {...process.env};
    });
    afterEach(function () {
      process.env = backupEnv;
    });
    it('should set right registry for npm', async function () {
      const config = {
        registry: 'my.registry',
        'strict-ssl': true,
        'package-lock': false
      };
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await setUpNpmConfig(nodePath, config);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should call npm install', async function () {
      const installSpyOn = jest.spyOn(npm, 'install');
      await installNpmDependencies(nodePath, {'mypackage': '1.2.3'});
      expect(installSpyOn.mock.calls[installSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use env var for registry', async function () {
      process.env.SAUCE_NPM_CACHE = 'npmland.io';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(runCfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use user registry', async function () {
      const cfg = _.cloneDeep(runCfg);
      cfg.npm ||= {};
      cfg.npm.registry = 'registryland.io';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(cfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use default registry', async function () {
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(runCfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use true as the default value for strictSSL', async function () {
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(runCfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use true as the default value for strictSSL if it\'s null in cfg', async function () {
      const cfg = _.cloneDeep(runCfg);
      cfg.npm ||= {};
      cfg.npm.strictSSL = null;
      cfg.npm.registry = 'test.strictSSL.null';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(runCfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should be able to set strictSSL to false', async function () {
      const cfg = _.cloneDeep(runCfg);
      cfg.npm ||= {};
      cfg.npm.strictSSL = false;
      cfg.npm.registry = 'test.strictSSL.false';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(cfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should be able to set strictSSL to true', async function () {
      const cfg = _.cloneDeep(runCfg);
      cfg.npm ||= {};
      cfg.npm.strictSSL = true;
      cfg.npm.registry = 'test.strictSSL.true';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(cfg, nodeCtx);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should be able to set cafile', async function () {
      const cfg = _.cloneDeep(runCfg);
      cfg.npm ||= {};
      cfg.npm.registry = 'test.cafile';
      process.env.CA_FILE = '/fake/path';
      const loadSpyOn = jest.spyOn(npm, 'configure');
      await prepareNpmEnv(cfg, nodePath);
      expect(loadSpyOn.mock.calls[loadSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
    it('should use rebuild node_modules', async function () {
      const rebuildSpyOn = jest.spyOn(npm, 'rebuild');
      const statSyncSpyOn = jest.spyOn(fs, 'statSync');
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      statSyncSpyOn.mockReturnValue({ isDirectory: () => true });
      await prepareNpmEnv(runCfg, nodeCtx);
      expect(rebuildSpyOn.mock.calls[rebuildSpyOn.mock.calls.length - 1]).toMatchSnapshot();

    });
    it('should use rebuild node_modules when not package installed', async function () {
      const cfg = _.cloneDeep(runCfg);
      delete cfg.npm;
      const rebuildSpyOn = jest.spyOn(npm, 'rebuild');
      const statSyncSpyOn = jest.spyOn(fs, 'statSync');
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      statSyncSpyOn.mockReturnValue({ isDirectory: () => true });
      await prepareNpmEnv(cfg, nodePath);
      expect(rebuildSpyOn.mock.calls[rebuildSpyOn.mock.calls.length - 1]).toMatchSnapshot();
    });
  });
  describe('.renameScreenshot', function () {
    it('replace path separator (backslash for Windows, forward slash for mac/linux) with __', function () {
      const spy = jest.spyOn(fs, 'renameSync').mockImplementation(function () { return undefined; });
      const nestedExample = path.join('nested', 'example.test.js');
      expect(renameScreenshot(nestedExample, 'old_path', 'new_path', 'screenshot.png')).toEqual(path.join('new_path', 'nested__example.test.js__screenshot.png'));
      expect(spy).toHaveBeenCalled();
    });
  });
  describe('.renameAsset', function () {
    it('root folder no need to rename asset with path separator', function () {
      const nestedExampleTest = path.join('assets', 'example.test.js.xml');
      expect(renameAsset({
        specFile: 'example.test.js.xml',
        oldFilePath: nestedExampleTest,
        resultsFolder: '/new_path'
      })).toEqual('assets/example.test.js.xml');
    });
    it('asset is in nested folder and replacing path separator with __', function () {
      const nestedExampleTest = path.join('nested', 'example.test.js.xml');
      const spy = jest.spyOn(fs, 'renameSync').mockImplementation(function () { return undefined;});
      expect(renameAsset({
        specFile: nestedExampleTest,
        oldFilePath: '/assets/example.test.js.xml',
        resultsFolder: '/new_path'
      })).toEqual(path.join(path.sep, 'new_path', 'nested__example.test.js.xml'));
      expect(spy).toHaveBeenCalled();
    });
  });
  describe('.getAbsolutePath', function () {
    it('returns absolute path unmodified', function () {
      jest.mock('fs');
      expect(getAbsolutePath('/absolute/path/to/asset/')).toEqual('/absolute/path/to/asset/');
    });
    it('translates relative path to absolute', function () {
      expect(getAbsolutePath('path/to/asset/')).toMatch(/\/path\/to\/asset\/$/);
    });
  });
  describe('.shouldRecordVideo', function () {
    let previousEnv: string | undefined;
    beforeEach(function () {
      previousEnv = process.env.SAUCE_CYPRESS_VIDEO_RECORDING;
    });
    afterEach(function () {
      process.env.SAUCE_CYPRESS_VIDEO_RECORDING = previousEnv;
    });
    it('returns true when SAUCE_CYPRESS_VIDEO_RECORDING is undefined', function () {
      expect(shouldRecordVideo()).toEqual(true);
    });
    it('returns false when SAUCE_CYPRESS_VIDEO_RECORDING is 0', function () {
      process.env.SAUCE_CYPRESS_VIDEO_RECORDING = '0';
      expect(shouldRecordVideo()).toEqual(false);
    });
    it('returns true when SAUCE_CYPRESS_VIDEO_RECORDING is 1', function () {
      process.env.SAUCE_CYPRESS_VIDEO_RECORDING = '1';
      expect(shouldRecordVideo()).toEqual(true);
    });
    it('returns true when SAUCE_CYPRESS_VIDEO_RECORDING is true', function () {
      process.env.SAUCE_CYPRESS_VIDEO_RECORDING = 'true';
      expect(shouldRecordVideo()).toEqual(true);
    });
    it('returns false when SAUCE_CYPRESS_VIDEO_RECORDING is false', function () {
      process.env.SAUCE_CYPRESS_VIDEO_RECORDING = 'false';
      expect(shouldRecordVideo()).toEqual(false);
    });
  });
  describe('.getArgs', function () {
    let backupArgv: string[];
    beforeEach(function () {
      backupArgv = process.argv;
      process.argv = [
        '/path/to/node',
        '/path/to/sauce-cypress-runner',
        '--suiteName', 'kitchen-sink-1',
        '--runCfgPath', './tests/kitchen-sink-tests/sauce-runner.json'
      ];
    });
    afterEach(function () {
      process.argv = backupArgv;
    });
    it('should parse the args', function () {
      const commandLineArgs = getArgs();
      expect(commandLineArgs).toMatchSnapshot();
      expect(getArgs()).toBe(commandLineArgs);
    });
  });
  describe('.getSuite', function () {
    it('should get a suite from a list', function () {
      const runCfg: SuitesContainer = {
        suites: [
          {name: 'hello'}
        ]
      };
      expect(getSuite(runCfg, 'hello')?.name).toEqual('hello');
      expect(getSuite(runCfg, 'non-existent')).toBeUndefined();
    });
  });
  describe('.getEnv', function () {
    let backupEnv: {[key: string]: string | undefined};
    beforeEach(function () {
      backupEnv = process.env;
      process.env = {
        'HELLO': 'WORLD',
        'FOO': 'BAR',
      };
    });
    afterEach(function () {
      process.env = backupEnv;
    });
    it('should parse env variables from runConfig', function () {
      const suite: Suite = {
        name: 'Demo suite',
        env: {
          'A': '1',
          'B': '2',
          'HELLO': '$HELLO',
        },
        config: {
          env: {
            'C': '3',
          }
        }
      };
      const env = getEnv(suite);
      expect(env).toMatchSnapshot();
      expect(env.FOO).toBeUndefined();
    });
  });
  describe('.escapeXML', function () {
    it('should escape special characters for XML', function () {
      const str = `<>&\\"`;
      const result = escapeXML(str);
      expect(result).toBe('&lt;&gt;&amp;\\&quot;');
    });
  });
});
