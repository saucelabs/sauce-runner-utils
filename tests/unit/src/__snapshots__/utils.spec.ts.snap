// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils .getArgs should parse the args 1`] = `
{
  "nodeBin": "/path/to/node",
  "runCfgPath": "./tests/kitchen-sink-tests/sauce-runner.json",
  "suiteName": "kitchen-sink-1",
}
`;

exports[`utils .getEnv should parse env variables from runConfig 1`] = `
{
  "A": "1",
  "B": "2",
  "C": "3",
  "HELLO": "WORLD",
}
`;

exports[`utils .prepareNpmEnv should be able to set strictSSL to false 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "test.strictSSL.false",
    "save": false,
    "strict-ssl": false,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should be able to set strictSSL to true 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "test.strictSSL.true",
    "save": false,
    "strict-ssl": true,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should call npm install 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  [
    "mypackage@1.2.3",
  ],
]
`;

exports[`utils .prepareNpmEnv should set right registry for npm 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "my.registry",
    "save": false,
    "strict-ssl": true,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should use default registry 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "https://registry.npmjs.org",
    "save": false,
    "strict-ssl": null,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should use env var for registry 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "npmland.io",
    "save": false,
    "strict-ssl": null,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should use rebuild node_modules 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  "--prefix",
  "/fake/runner",
]
`;

exports[`utils .prepareNpmEnv should use rebuild node_modules when not package installed 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  "--prefix",
  "/fake/runner",
]
`;

exports[`utils .prepareNpmEnv should use true as the default value for strictSSL 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "https://registry.npmjs.org",
    "save": false,
    "strict-ssl": null,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should use true as the default value for strictSSL if it's null in cfg 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "https://registry.npmjs.org",
    "save": false,
    "strict-ssl": null,
    "update-notifier": false,
  },
]
`;

exports[`utils .prepareNpmEnv should use user registry 1`] = `
[
  {
    "nodePath": "node-bin",
    "npmPath": "npm-bin",
    "useGlobals": false,
  },
  {
    "audit": false,
    "fund": false,
    "json": false,
    "legacy-peer-deps": true,
    "noproxy": "registry.npmjs.org",
    "package-lock": false,
    "registry": "registryland.io",
    "save": false,
    "strict-ssl": null,
    "update-notifier": false,
  },
]
`;
